lib/epochs.js

```diff
@@ -20,109 +20,211 @@ const {
-        })
-
-      cb(null, epochs)
-    })
+  const allGetters = {
+    author (epochRoot, cb) {
+      ssb.metafeeds.findRootFeedId(epochRoot.value.author, cb)
+    },
+    epochKey (epochRoot, cb) {
+      cb(null, epochRoot.value.content.groupKey)
+    },
+    members (epochRoot, cb) {
+      getMembers(ssb, epochRoot.key, cb)
+    }
   }
 
-  function getMembers (epochRoot, cb) {
-    if (cb === undefined) return p(getMembers)(epochRoot)
+  return {
+    getEpochs (groupId, cb) {
+      if (cb === undefined) return p(this.getEpochs)(groupId)
 
-    getEpochMembers(ssb, epochRoot, cb)
-  }
+      const getters = pluck(allGetters, ['author', 'epochKey'])
+      reduceEpochs(ssb, groupId, getters, (err, reduce) => {
+        if (err) return cb(clarify(err, 'Failed to resolve epoch @tangle/reduce'))
 
-  return {
-    getEpochs,
-    getMembers,
+        const epochs = reduce.graph.connectedNodes
+          .map(node => {
+            const info = {
+              id: node.key, // alias: epochRootId
+              previous: node.previous,
+              ...node.data[node.key]
+            }
+            info.epochKey = Buffer.from(info.epochKey, 'base64')
+            return info
+          })
+
+        cb(null, epochs)
+      })
+    },
+
+    getMembers (epochRootId, cb) {
+      if (cb === undefined) return p(this.getMembers)(epochRootId)
+
+      getMembers(ssb, epochRootId, cb)
+    },
+
+    getMissingMembers (groupId, cb) {
+      if (cb === undefined) return p(this.getMissingMembers)(groupId)
+
+      reduceEpochs(ssb, groupId, allGetters, (err, reduce) => {
+        if (err) return cb(clarify(err, 'Failed to resolve epoch @tangle/reduce'))
+
+        if (reduce.graph.connectedNodes.length === 1) return cb(null, [])
+        // INFO if there is only one connectedNode, there is only a
+        // single epoch, so there's no need to check anything
+
+        const result = reduce.graph.connectedNodes.reduce((acc, node) => {
+          // For each node (epoch) in the tangle, determine which feeds have
+          // historically been added / excluded.
+          const addedSoFar = new Set()
+          const excludedSoFar = new Set()
+          ;[node.key, ...reduce.graph.getHistory(node.key)]
+            .map(key => reduce.graph.getNode(key))
+            .forEach(historyNode => {
+              const { added, toExclude } = historyNode.data[historyNode.key].members
+              added.forEach(feedId => addedSoFar.add(feedId))
+
+              if (historyNode.key === node.key) return
+              // INFO node members toExclude is talking about what should happen
+              // *after* it in the graph, so if we're enquiring about a
+              // particular epoch, we don't include that epoch's toExclude
+              toExclude.forEach(feedId => excludedSoFar.add(feedId))
+            })
+
+          // Check if those who should be present are present
+          const shouldBePresent = difference(addedSoFar, excludedSoFar)
```

Comment:
Doesn't this ignore people who have been removed and added back?

Reply:
yeah we're not covering that case for the moment. LATER! (/never)